src/neml2/models/Model.cxx

```diff
@@ -32,45 +32,254 @@ ParameterSet
-      std::vector<std::vector<std::string>>();
+  params.set<std::vector<LabeledAxisAccessor>>("additional_outputs");
+  params.set<bool>("use_AD_first_derivative") = false;
+  params.set<bool>("use_AD_second_derivative") = false;
   return params;
 }
 
 Model::Model(const ParameterSet & params)
   : NEML2Object(params),
     _input(declareAxis()),
-    _output(declareAxis())
+    _output(declareAxis()),
+    _AD_1st_deriv(params.get<bool>("use_AD_first_derivative")),
+    _AD_2nd_deriv(params.get<bool>("use_AD_second_derivative"))
 {
-  for (const auto & var : params.get<std::vector<std::vector<std::string>>>("additional_outputs"))
-    _additional_outputs.insert(LabeledAxisAccessor{var});
+  check_AD_limitation();
+  for (const auto & var : params.get<std::vector<LabeledAxisAccessor>>("additional_outputs"))
+    _additional_outputs.insert(var);
   setup();
 }
 
+void
+Model::check_AD_limitation() const
+{
+  // AD_1st_deriv   AD_2nd_deriv   comment
+  // true           true           okay, just slow
+  // true           false          error, this is a weird case
+  // false          true           okay
+  // false          false          great, everything handcoded
+  if (_AD_1st_deriv && !_AD_2nd_deriv)
+    throw NEMLException("AD derivative is requested, but AD second derivative is not requested.");
+}
+
+void
+Model::use_AD_derivatives(bool first, bool second)
+{
+  _AD_1st_deriv = first;
+  _AD_2nd_deriv = second;
+  check_AD_limitation();
+}
+
+void
+Model::trace_parameters(const std::map<std::string, bool> & params)
+{
+  const auto & model_params = named_parameters(true);
+  for (auto && [name, requires_grad] : params)
+    model_params[name].requires_grad_(requires_grad);
+}
+
+void
+Model::set_parameters(const std::map<std::string, torch::Tensor> & params)
+{
+  const auto & model_params = named_parameters(true);
+  for (auto && [name, val] : params)
+  {
+    auto req_grad = model_params[name].requires_grad();
+    model_params[name].requires_grad_(false);
+    model_params[name].copy_(val);
+    model_params[name].requires_grad_(req_grad);
+  }
+}
+
+BatchTensor<1>
+Model::dparam(const LabeledVector & out, const std::string & param) const
+{
+  const auto & outval = out.tensor();
+  BatchTensor<1> pval = named_parameters(true)[param];
+
+  BatchTensor<1> dout_dp(
+      out.batch_size(), utils::add_shapes(outval.base_sizes(), pval.base_sizes()), out.options());
+
+  if (pval.batch_sizes()[0] == 1)
+  {
+    for (TorchSize b = 0; b < outval.batch_sizes()[0]; b++)
+      for (TorchSize i = 0; i < outval.base_sizes()[0]; i++)
+      {
+        BatchTensor<1> grad_outputs = torch::zeros_like(outval.index({b}));
+        grad_outputs.index_put_({i}, 1.0);
+        try
+        {
+          auto jac_row =
+              torch::autograd::grad({outval.index({b})}, {pval}, {grad_outputs}, true)[0];
+          dout_dp.index_put_({b, i, torch::indexing::Slice()}, jac_row);
+        }
+        catch (c10::Error &)
```

Comment:
Fix this either by:

1. Figuring out the right torch options to grad...
2. Detecting the issue with logic in the loop versus catching here
3. Worst case: check error string...

Reply:
So it turns out we need the right options (allow_unused = true) plus a few other checks. But ya, we can get rid of the try catch.